src/sentry/search/snuba/backend.py

```diff
@@ -49,6 +51,57 @@ def unassigned_filter(unassigned, projects):
     return query
 
 
+def linked_filter(linked, projects):
+    """
+    Builds a filter for whether or not a Group has an issue linked via either
+    a PlatformExternalIssue or an ExternalIssue.
+    """
+    # Note that there is no filter being applied here with `projects`. We rely on
+    # the planner to avoid a seq scan on the PlatformExternalIssue table via
+    # the more obvious semi-join from Group to PlatformExternalIssue. We notably
+    # do not attempt to filter on `projects` as the the path from a
+    # PlatformExternalIssue to the Project.id goes through the Group table, and
+    # thus would require a seq scan on the PlatformExternalIssue table or an index
+    # scan over the Group rows for the given projects, which would negate the
+    # benefits of any indexes used for other filters. This would be more reasonable if
+    # we had a redundant `project_id` (like GroupLink.project_id) on this table.
+    group_ids_with_platform_external_issue = PlatformExternalIssue.objects.values_list(
+        "group_id", flat=True
+    )
+
+    # NB: The join ExternalIssue.id=GroupLink.linked_id is not needed, since we can
+    # just discriminate on GroupLink.linked_type.
+    group_ids_with_external_issue = GroupLink.objects.filter(
+        project_id__in=[p.id for p in projects],
+        linked_type=GroupLink.LinkedType.issue,
+        relationship=GroupLink.Relationship.references,
+    ).values_list("group_id", flat=True)
+
+    # The naive way, implemented here, combines the two subexpressions with an OR.
+    # This may notably introduce suboptimal query plans due to two issues:
+    #   - There are no statistics on the overlap of the two subexpressions
+    #   - The planner might not realize it can use an optimized semi-join
+    # It's possible that these suboptimalities might not be the bottleneck for perf
+    # in production. But if they are, we may want to optimize for the happy path
```

Comment:
@iProgramStuff : I unfortunately can't test this myself very easily since I don't have access to a representative database 😅 . Would it be possible to run some queries against the production sentry.io postgres database to test performance? 

For the following queries, I've used `project_id IN (111)`, but it would be good to repeat the following queries for 3 projects, where each fall into one of 3 categories respectively:
- Has many groups that are linked to ExternalIssue but none to PlatformExternalIssue.
- Has many groups that are linked to PlatformExternalIssue but none to ExternalIssue.
- Has groups that are linked to a mix of ExternalIssue and PlatformExternalIssue.

## OR query
### From PR
```sql
-- OR QUERY
-- from pr
EXPLAIN ANALYZE SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND ("sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_grouplink" U0 WHERE (
            U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2
        )
        ) OR "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1 FROM "sentry_platformexternalissue" U0
        )
    )
)
LIMIT 5001;
```
### Alternative query with group filter
```sql
SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND (
        "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2)
        )
        OR "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE V0."group_id" IN (
                SELECT U0."id" AS Col1
                FROM "sentry_groupedmessage" U0 WHERE U0."project_id" IN (111)
            )
        )
    )
) LIMIT 5001;
```
 
## PlatformExternalIssue happy path
### Integration query
```sql
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   JOIN sentry_groupedmessage ON
        sentry_groupedmessage.id = sentry_platformexternalissue.group_id
   WHERE sentry_groupedmessage.project_id IN (111)
)
```

### Issue query
```sql
SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_platformexternalissue" U0
    )
)
LIMIT 5001;
```

## ExternalIssue happy path
### Integration query
```sql
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (111)
)
```

### Issue query
```sql
EXPLAIN ANALYZE SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_grouplink" U0 WHERE (
            U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2
        )
    )
)
LIMIT 5001;
```

Reply:
Maybe we can limit these filters behind a feature flag first to test it out?

Reply:
Seems reasonable to me. I can add an OrganizationFeature called `organizations:is-linked-search` ?
Though, I think it'll still be helpful to run these `EXPLAIN ANALYZE`s to confirm the performance at various corner cases

Reply:
@wedamija : An alternative code that I was alluding to is:

```py
platform_qs = PlatformExternalIssue.objects.filter(project_id__in=[p.id for p in projects])
integration_qs = GroupLink.objects.filter(
    project_id__in=[p.id for p in projects],
    linked_type=GroupLink.LinkedType.issue,
    relationship=GroupLink.Relationship.references,
)

group_linked_to_platform_issue_q = Q(id__in=platform_qs.values_list("group_id", flat=True))
group_linked_to_integration_issue_q = Q(id__in=integration_qs.values_list("group_id", flat=True))

# Usually a user will either only have PlatformExternalIssues or only have ExternalIssues,
# i.e. in most cases, at most one of the below expressions evaluates to True:
platform_issue_exists = platform_qs.exists()
integration_issue_exists = integration_qs.exists()
# By optimizing for this case, we're able to produce a filter that roughly translates to
# `WHERE group_id IN (SELECT group_id FROM one_issue_table WHERE ...)`, which the planner
# is able to optimize with the semi-join strategy.
if platform_issue_exists and not integration_issue_exists:
    query = group_linked_to_platform_issue_q
elif integration_issue_exists and not platform_issue_exists:
    query = group_linked_to_integration_issue_q
# ...but if we don't have exactly one type of issues, fallback to doing the OR.
else:
    query = group_linked_to_platform_issue_q | group_linked_to_integration_issue_q
```

Getting the times / plans for the following 4 queries would be useful to confirm if this alternative strategy is more performant.

```sql
-- Current query that does an OR
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (<redacted>)
    AND (
        "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
        )
        OR "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE (V0."project_id" IN (<redacted>) 
        )
    )
) LIMIT 5001; 
```

vs.

A
```sql
-- Existence of integration external issue
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (111) AND linked_type = 3 AND relationship = 2
)
```

B
```sql
-- Existence of platform external issue
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   WHERE sentry_platformexternalissue.project_id IN (111)
)
```

C
```
-- Happypath for integration external issue
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )
) LIMIT 5001; 
```
D
```
-- Happy path for platform external issue
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE (V0."project_id" IN (111))
    )
) LIMIT 5001; 
```

The alternative code I suggested I think will be sensible, if we see that `A + B + C < X` and `A + B + D < X`.

The non-happy-path case takes time `A+B+C+D` which I wouldn't be too concerned about if it's more than `X`, but guessing it's likely a neglibible difference.

Reply:
@saifelse I think this is a good suggestion. With the OR in place we only filter `groupedmessage` via the project id, which ends up really slow. This plan is from a subsequent run, but initially this took 13 seconds:

X
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1
        FROM "sentry_grouplink" U0
        WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )
    OR 
    "sentry_groupedmessage"."id" IN (
        SELECT V0."group_id" AS Col1
        FROM "sentry_platformexternalissue" V0
        WHERE (V0."project_id" IN (<redacted>))
    )
)
LIMIT 5001;

Limit (cost=994.88..2613.87 rows=5001 width=8) (actual time=4.004..1336.584 rows=426 loops=1)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=994.88..419233.21 rows=1291921 width=8) (actual time=4.003..1336.416 rows=426 loops=1)
Index Cond: (project_id = <redacted>)
Filter: ((hashed SubPlan 1) OR (hashed SubPlan 2))
Rows Removed by Filter: 1836401
Heap Fetches: 266499
SubPlan 1
-> Bitmap Heap Scan on sentry_grouplink u0 (cost=10.00..586.61 rows=346 width=8) (actual time=0.152..2.576 rows=441 loops=1)
Recheck Cond: (project_id = <redacted>)
Filter: ((linked_type = 3) AND (relationship = 2))
Heap Blocks: exact=301
-> Bitmap Index Scan on sentry_grouplink_project_id (cost=0.00..9.91 rows=466 width=0) (actual time=0.094..0.094 rows=441 loops=1)
Index Cond: (project_id = <redacted>)
SubPlan 2
-> Bitmap Heap Scan on sentry_platformexternalissue v0 (cost=7.28..405.86 rows=386 width=8) (actual time=0.105..1.095 rows=422 loops=1)
Recheck Cond: (project_id = <redacted>)
Heap Blocks: exact=167
-> Bitmap Index Scan on sentry_platformexternalissue_project_id_ff4e3d10 (cost=0.00..7.18 rows=386 width=0) (actual time=0.080..0.080 rows=422 loops=1)
Index Cond: (project_id = <redacted>)
Planning time: 3.096 ms
Execution time: 1336.923 ms
```

VS doing just one statement at a time, we actually and up searching the index with `project_id` and `group_id`, much nicer:
C
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1
        FROM "sentry_grouplink" U0
        WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )

)
LIMIT 5001;

Limit (cost=588.05..1564.54 rows=3 width=8) (actual time=3.359..18.947 rows=426 loops=1)
-> Nested Loop (cost=588.05..1564.54 rows=3 width=8) (actual time=3.356..18.880 rows=426 loops=1)
-> HashAggregate (cost=587.48..590.94 rows=346 width=8) (actual time=3.193..3.348 rows=426 loops=1)
Group Key: u0.group_id
-> Bitmap Heap Scan on sentry_grouplink u0 (cost=10.00..586.61 rows=346 width=8) (actual time=0.332..3.060 rows=441 loops=1)
Recheck Cond: (project_id = <redacted>)
Filter: ((linked_type = 3) AND (relationship = 2))
Heap Blocks: exact=301
-> Bitmap Index Scan on sentry_grouplink_project_id (cost=0.00..9.91 rows=466 width=0) (actual time=0.285..0.285 rows=441 loops=1)
Index Cond: (project_id = <redacted>)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=0.57..2.80 rows=1 width=8) (actual time=0.026..0.036 rows=1 loops=426)
Index Cond: ((project_id = <redacted>) AND (id = u0.group_id))
Heap Fetches: 5895
Planning time: 7.304 ms
Execution time: 19.184 ms
```
and
D
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT V0."group_id" AS Col1
        FROM "sentry_platformexternalissue" V0
        WHERE (V0."project_id" IN (<redacted>))
    )
)
LIMIT 5001;

Limit (cost=407.40..1496.29 rows=4 width=8) (actual time=1.448..12.650 rows=422 loops=1)
-> Nested Loop (cost=407.40..1496.29 rows=4 width=8) (actual time=1.447..12.599 rows=422 loops=1)
-> HashAggregate (cost=406.83..410.69 rows=386 width=8) (actual time=1.269..1.346 rows=422 loops=1)
Group Key: v0.group_id
-> Bitmap Heap Scan on sentry_platformexternalissue v0 (cost=7.28..405.86 rows=386 width=8) (actual time=0.151..1.182 rows=422 loops=1)
Recheck Cond: (project_id = <redacted>)
Heap Blocks: exact=167
-> Bitmap Index Scan on sentry_platformexternalissue_project_id_ff4e3d10 (cost=0.00..7.18 rows=386 width=0) (actual time=0.130..0.130 rows=422 loops=1)
Index Cond: (project_id = <redacted>)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=0.57..2.80 rows=1 width=8) (actual time=0.019..0.026 rows=1 loops=422)
Index Cond: ((project_id = <redacted>) AND (id = v0.group_id))
Heap Fetches: 5953
Planning time: 4.029 ms
Execution time: 12.799 ms
```

Even with the extra IO wait time from the extra queries I think this is a better option. Just for completeness, here are the query plans for A and B as well.
A
```
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (<redacted>) AND linked_type = 3 AND relationship = 2
)

Result (cost=2.73..2.74 rows=1 width=1) (actual time=0.088..0.088 rows=1 loops=1)
InitPlan 1 (returns $0)
-> Index Scan using sentry_grouplink_project_id on sentry_grouplink (cost=0.42..799.11 rows=346 width=0) (actual time=0.087..0.087 rows=1 loops=1)
Index Cond: (project_id = 2277697)
Filter: ((linked_type = 3) AND (relationship = 2))
Planning time: 1.146 ms
Execution time: 0.121 ms
```

B
```
explain analyze SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   WHERE sentry_platformexternalissue.project_id IN (<redacted>)
)

Result (cost=0.92..0.93 rows=1 width=1) (actual time=8.910..8.911 rows=1 loops=1)
InitPlan 1 (returns $0)
-> Index Only Scan using sentry_platformexternalissue_project_id_ff4e3d10 on sentry_platformexternalissue (cost=0.29..243.02 rows=386 width=0) (actual time=8.906..8.906 rows=1 loops=1)
Index Cond: (project_id = 2277697)
Heap Fetches: 1
Planning time: 12.794 ms
Execution time: 9.816 ms
```

Reply:
@wedamija : applied the changes!